cli/cmd/check.go

```diff
@@ -254,103 +255,75 @@ func runExtensionChecks(cmd *cobra.Command, wout io.Writer, werr io.Writer, opts
-			// create a healthchecker that checks if the namespace exists
-			categoryID := healthcheck.CategoryID(ns.Labels[k8s.LinkerdExtensionLabel])
-			checkers := []healthcheck.Checker{}
-			checkers = append(checkers,
-				*healthcheck.NewChecker(fmt.Sprintf("%s extension Namespace exists", categoryID)).
-					WithHintAnchor(fmt.Sprintf("%s-ns-exists", categoryID)).
-					Fatal().
-					WithCheck(func(ctx context.Context) error {
-						_, err := kubeAPI.GetNamespaceWithExtensionLabel(ctx, string(categoryID))
-						if err != nil {
-							return err
-						}
-						return nil
-					}))
-			hc := healthcheck.NewHealthChecker([]healthcheck.CategoryID{categoryID},
-				&healthcheck.Options{
-					ControlPlaneNamespace: controlPlaneNamespace,
-					CNINamespace:          cniNamespace,
-					DataPlaneNamespace:    opts.namespace,
-					KubeConfig:            kubeconfigPath,
-					KubeContext:           kubeContext,
-					Impersonate:           impersonate,
-					ImpersonateGroup:      impersonateGroup,
-					APIAddr:               apiAddr,
-					VersionOverride:       opts.versionOverride,
-					RetryDeadline:         time.Now().Add(opts.wait),
-					CNIEnabled:            opts.cniEnabled,
-				})
-			hc.AppendCategories(*healthcheck.NewCategory(categoryID, checkers, true))
-			success := healthcheck.RunChecks(wout, werr, hc, opts.output)
-			if !success {
-				os.Exit(1)
+			extensionCmd := fmt.Sprintf("linkerd-%s", extension)
+			path, err = exec.LookPath(extensionCmd)
+			results.Results = []healthcheck.CheckResult{
+				{
+					Category:    healthcheck.CategoryID(extensionCmd),
+					Description: fmt.Sprintf("Linkerd extension command %s exists", extensionCmd),
+					Err:         err,
+					HintAnchor:  "extensions",
+					Warning:     true,
+				},
 			}
 		}
-		if err != nil {
-			return err
-		}
-	}
-	return nil
-}
+		if err == nil {
+			plugin := exec.Command(path, args...)
+			output, _ := plugin.Output()
+			extensionResults, err := healthcheck.ParseJSONCheckOutput(output)
+			if err != nil {
 
-func getGlobalFlags(lf *pflag.FlagSet) []string {
-	cmdLineFlags := []string{}
-	lf.VisitAll(func(f *pflag.Flag) {
-		val := f.Value.String()
-		if val != "" {
-			cmdLineFlags = append(cmdLineFlags, fmt.Sprintf("--%s=%s", f.Name, val))
+				return false, fmt.Errorf("invalid extension check output (JSON object expected):\n%s\n[%s]", string(output), err)
+			}
+			results.Results = append(results.Results, extensionResults.Results...)
 		}
-	})
-
-	return cmdLineFlags
+		extensionSuccess := healthcheck.RunChecks(wout, werr, results, opts.output)
+		if !extensionSuccess {
+			success = false
+		}
+	}
+	return success, nil
 }
 
-func getSharedFlags(lf *pflag.FlagSet, lf2 *pflag.FlagSet) []string {
+func getExtensionCheckFlags(lf *pflag.FlagSet) []string {
+	extensionFlags := []string{
+		"api-addr", "context", "as", "as-group", "kubeconfig", "linkerd-namespace", "verbose",
+		"namespace", "proxy", "wait",
```

Comment:
Not all extensions accept the same flags, and sending an unknown flag will cause an error. For example the mc check call is failing because it doesn't recognize `--proxy`.